drivers/can/can_nxp_s32_canxl.c

```diff
@@ -0,0 +1,1020 @@
+				frame->id, (uint8_t *)&frame->data, &data->tx_msg[alloc]);
+	config->irq_enable_func();
+	k_mutex_unlock(&data->tx_mutex);
+
+	if (status != CANEXCEL_STATUS_SUCCESS) {
+		return -EIO;
+	}
+
+	return 0;
+}
+
+static void nxp_s32_zcan_timming_to_canxl_timming(const struct can_timing *timing,
+							Canexcel_Ip_TimeSegmentType *canxl_timing)
+{
+	LOG_DBG("propSeg: %d, phase_seg1: %d, phase_seg2: %d, prescaler: %d, sjw: %d",
+		timing->prop_seg, timing->phase_seg1, timing->phase_seg2,
+		timing->prescaler, timing->sjw);
+
+	canxl_timing->propSeg = timing->prop_seg - 1U;
+	canxl_timing->phaseSeg1 = timing->phase_seg1 - 1U;
+	canxl_timing->phaseSeg2 = timing->phase_seg2 - 1U;
+	canxl_timing->preDivider = timing->prescaler - 1U;
+	canxl_timing->rJumpwidth = timing->sjw - 1U;
+}
+
+static int can_nxp_s32_set_timing(const struct device *dev,
+			const struct can_timing *timing)
+{
+	const struct can_nxp_s32_config *config = dev->config;
+	struct can_nxp_s32_data *data = dev->data;
+	Canexcel_Ip_TimeSegmentType can_time_segment = {0};
+
+	if (data->started) {
+		return -EBUSY;
+	}
+
+	LOG_DBG("Setting CAN bitrate %d:", config->bitrate);
+	nxp_s32_zcan_timming_to_canxl_timming(timing, &can_time_segment);
+
+	/* Set timing for CAN instance*/
+	CanXL_SetBaudRate(config->base_sic, &can_time_segment);
+
+	return 0;
+}
+
+#ifdef CONFIG_CAN_FD_MODE
+static int can_nxp_s32_set_timing_data(const struct device *dev,
+			const struct can_timing *timing_data)
+{
+	const struct can_nxp_s32_config *config = dev->config;
+	struct can_nxp_s32_data *data = dev->data;
+	Canexcel_Ip_TimeSegmentType can_fd_time_segment = {0};
+
+	if (data->started) {
+		return -EBUSY;
+	}
+
+	LOG_DBG("Setting CAN-FD bitrate %d:", config->bitrate_data);
+	nxp_s32_zcan_timming_to_canxl_timming(timing_data, &can_fd_time_segment);
+
+	/* Set timing for CAN FD instance*/
+	CanXL_SetFDBaudRate(config->base_sic, &can_fd_time_segment);
+
+	return 0;
+}
+#endif
+
+static void can_nxp_s32_err_callback(const struct device *dev,
+					Canexcel_Ip_EventType eventType,
+					uint32 u32SysStatus,
+					const Canexcel_Ip_StateType *canexcelState)
+{
+	const struct can_nxp_s32_config *config = dev->config;
+	struct can_nxp_s32_data *data = dev->data;
+	enum can_state state;
+	struct can_bus_err_cnt err_cnt;
+	void *cb_data = data->state_change_cb_data;
+	can_tx_callback_t function;
+	int alloc;
+	void *arg;
+
+	switch (eventType) {
+	case CANEXCEL_EVENT_TX_WARNING:
+		LOG_DBG("Tx Warning (error 0x%x)", u32SysStatus);
+		break;
+	case CANEXCEL_EVENT_RX_WARNING:
+		LOG_DBG("Rx Warning (error 0x%x)", u32SysStatus);
+		break;
+	case CANEXCEL_EVENT_BUSOFF:
+		LOG_DBG("Bus Off (error 0x%x)", u32SysStatus);
+		break;
+	case CANEXCEL_EVENT_ERROR:
+		LOG_DBG("Error Format Frames (error 0x%x)", u32SysStatus);
+		break;
+	case CANEXCEL_EVENT_ERROR_FD:
+		LOG_DBG("Error Data Phase (error 0x%x)", u32SysStatus);
+		break;
+	case CANEXCEL_EVENT_PASSIVE:
+		LOG_DBG("Error Passive (error 0x%x)", u32SysStatus);
+		break;
```

Comment:
Could this be used for implementing CAN stats support?

Reply:
I added support stats for them

Reply:
I left out the support https://github.com/zephyrproject-rtos/zephyr/pull/52450#discussion_r1178313786
so can't use the current implementing CAN stats support